Helm: remove config init container when there are no extra configuration sources

atykhyy opened the pull request:
This PR adds a bit of helm chart logic to remove the build-config init container when the `daemon.configSources` helm value is set to `config-map:cilium-config` (i.e. there are no extra configuration sources to read) and restores the old daemon behavior (reading configuration straight from the cilium-config configmap mounted into the agent container).

Rationale: not everyone needs it, and it's one more moving part to keep track of in an already extremely complicated piece of software.

```release-note
Disables the configuration resolver InitContainer when CiliumNodeConfig is not a configuration source.
```

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -806,6 +806,14 @@ daemon:
+  # configuration of the Cilium daemon.
+  #
+  # By default, this functionality is enabled.
+  enablePerNodeConfiguration: true
```
This is already configurable via the Helm value `daemon.configSources`. If it is exactly equal to `config-map:cilium-config`, then it is safe to disable the config resolver. Perhaps that is simpler than adding another knob.

Separately, if we are not allowing for dynamic configuration, then we should consider not creating the CiliumNodeConfig CRD.

atykhyy replied:
That's a neat idea!

I don't know about the CRD, though. If users change the value of `daemon.configSources` with `helm upgrade` to something other than `config-map:cilium-config`, would helm know to install the CRD?

squeed replied:
Helm doesn't create the CRD, the operator does. So, it should eventually wake up and create the CRD.

That said, we don't currently conditionally create CRDs, so I suspect this is not worth the bother.

tommyp1ckles commented on install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:
```diff
@@ -10,6 +10,7 @@
 
 {{- $kubeProxyReplacement := (coalesce .Values.kubeProxyReplacement "false") -}}
 {{- $envoyDS := eq (include "envoyDaemonSetEnabled" .) "true" -}}
+{{- $buildDaemonConfig := not (eq .Values.daemon.configSources "config-map:cilium-config") -}}
```
Wouldn't we also want to omit the initContainer in the case that we're only using a config-map, but it's being configured to be something other than `cilium-config`, for ex: `--values daemon.configSources=config-map:some-config`?

atykhyy replied:
Sure, it's easy, and would be consistent with what the build-config init container does. But the configmap name `cilium-config` is hard-coded everywhere else in the chart templates (cilium-operator, cilium-preflight, clustermesh-apiserver), so I am not sure what that changing just the daemon's configmap would do.

tommyp1ckles reviewed (approved):
Looks good, tested the regex and it successfully omits the config initContainer when the singular source is not a config-map.

atykhyy committed:
Remove config init container when there are no extra sources

This commit adds a bit of helm chart logic to remove the build-config
init container when the `daemon.configSources` helm value is set to
a single configmap, e.g. `config-map:cilium-config` or `config-map:abc`
(i.e. there are no extra configuration sources to read) and restores
the old agent daemonset behavior (reading configuration straight from
the configmap mounted into the agent container).

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>
